### PR TITLE
1604 custom pba

### PR DESCRIPTION
**Pull request by VakarisZ**
# What does this PR do?

Includes custom PBA into agent and puppet 

Add any further explanations here.

## PR Checklist
* [ ] Have you added an explanation of what your changes do and why you'd like to include them?
* [ ] Is the TravisCI build passing?
* [ ] Was the CHANGELOG.md updated to reflect the changes?
* [ ] Was the documentation framework updated to reflect the changes?

## Testing Checklist

* [ ] Added relevant unit tests?
* [x] Have you successfully tested your changes locally? Elaborate:
    > Tested by running on windows, checking attack and security reports
* [ ] If applicable, add screenshots or log transcripts of the feature working

## Explain Changes

Are the commit messages enough? If not, elaborate.

### CODE REVIEW

File: monkey/infection_monkey/post_breach/custom_pba/users_custom_pba.py [hunk 1]
```diff
@@ -23,45 +24,45 @@ class UsersPBA(PBA):
...
+    def should_run(options):
         if not is_windows_os():
-            if WormConfiguration.PBA_linux_filename or WormConfiguration.custom_PBA_linux_cmd:
+            if options["linux_filename"] or options["linux_command"]:
                 return True
         else:
-            if WormConfiguration.PBA_windows_filename or WormConfiguration.custom_PBA_windows_cmd:
+            if options["windows_filename"] or options["windows_command"]:
                 return True
         return False
```
Comment (shreyamalviya) [inline monkey/infection_monkey/post_breach/custom_pba/users_custom_pba.py:67]: Consider switching the `if` conditions here too in the way you've done it in `__init__()`.

File: monkey/infection_monkey/master/automated_master.py [hunk 2]
```diff
@@ -213,6 +214,15 @@ def _run_payload(self, payload: Tuple[str, Dict]):
...
 
+    def _run_PBAs(
+        self, plugins: Iterable[Any], callback: Callable[[Any], None], custom_pba_options: Mapping
+    ):
+        self._run_plugins(plugins, "post-breach action", callback)
+
+        command, result = self._puppet.run_custom_pba(custom_pba_options)
+        telem = PostBreachTelem(POST_BREACH_FILE_EXECUTION, command, result)
+        self._telemetry_messenger.send_telemetry(telem)
+
```
Comment (shreyamalviya) [inline monkey/infection_monkey/master/automated_master.py:225]:
```suggestion
    def _run_PBAs(
        self, plugins: Iterable[Any], callback: Callable[[Any], None], custom_pba_options: Mapping
    ):
        self._run_plugins(plugins, "post-breach action", callback)

        command, result = self._puppet.run_custom_pba(custom_pba_options)
        telem = PostBreachData(POST_BREACH_FILE_EXECUTION, command, result)
        self._telemetry_messenger.send_telemetry(telem)

```

File: monkey/infection_monkey/master/automated_master.py [hunk 3]
```diff
@@ -15,6 +15,7 @@
 from infection_monkey.utils.threading import create_daemon_thread, interruptible_iter
 from infection_monkey.utils.timer import Timer
 
+from ..post_breach.custom_pba.custom_pba import CustomPBA
```
Comment (ilija-lazoroski) [inline monkey/infection_monkey/master/automated_master.py:18]: Add `CustomPBA` in `__init__` and use it from there. Just for readability and consistency.

File: monkey/tests/unit_tests/infection_monkey/post_breach/actions/test_users_custom_pba.py [hunk 4]
```diff
@@ -2,7 +2,7 @@
 
 import pytest
 
-from infection_monkey.post_breach.actions.users_custom_pba import UsersPBA
+from infection_monkey.post_breach.custom_pba.users_custom_pba import UsersPBA
```
Comment (ilija-lazoroski) [inline monkey/tests/unit_tests/infection_monkey/post_breach/actions/test_users_custom_pba.py:5]:
```suggestion
from infection_monkey.post_breach.custom_pba.custom_pba import CustomPBA
```

**Review by ilija-lazoroski — CHANGES_REQUESTED** [review]
Rebase, fix conflicts,unit tests and retest. Nice job though!

File: monkey/infection_monkey/post_breach/custom_pba/custom_pba.py [hunk 5]
```diff
@@ -19,50 +20,58 @@
...
     @staticmethod
-    def should_run(class_name):
+    def should_run(options):
         if not is_windows_os():
-            if WormConfiguration.PBA_linux_filename or WormConfiguration.custom_PBA_linux_cmd:
+            if options["linux_filename"] or options["linux_command"]:
                 return True
         else:
-            if WormConfiguration.PBA_windows_filename or WormConfiguration.custom_PBA_windows_cmd:
+            if options["windows_filename"] or options["windows_command"]:
```
Comment (mssalvatore) [inline monkey/infection_monkey/post_breach/custom_pba/custom_pba.py:74]: Move to `AutomatedMaster`

File: monkey/infection_monkey/master/automated_master.py [hunk 6]
```diff
@@ -212,6 +209,14 @@ def _run_payload(self, payload: Tuple[str, Dict]):
 
         self._puppet.run_payload(name, options, self._stop)
 
+    def _run_pbas(
+        self, plugins: Iterable[Any], callback: Callable[[Any], None], custom_pba_options: Mapping
+    ):
+        self._run_plugins(plugins, "post-breach action", callback)
+
+        if CustomPBA.should_run(custom_pba_options):
+            self._run_plugins([("CustomPBA", custom_pba_options)], "post-breach action", callback)
```
Comment (mssalvatore) [inline monkey/infection_monkey/master/automated_master.py:218]: Try to find a way to avoid coupling the Master to any plugins.

**Review by mssalvatore — APPROVED** [review]

**Pull request merged**